extensions/lifecycle/tasks/LifecycleTask.js

```diff
@@ -534,6 +536,100 @@ class LifecycleTask extends BackbeatTask {
         });
     }
 
+    /**
+     * check if rule applies for a given date or calculed days.
+     * @param {array} rule - bucket lifecycle rule
+     * @param {number} daysSinceInitiated - Days passed since entity (object or version) last modified
+     * NOTE: entity is not an in-progress MPU or a delete marker.
+     * @param {number} currentDate - current date
+     * @return {boolean} true if rule applies - false otherwise.
+     */
+    _isRuleApplying(rule, daysSinceInitiated, currentDate) {
+        if (rule.Expiration && this._supportedRules.includes('expiration')) {
+            if (rule.Expiration.Days !== undefined && daysSinceInitiated >= rule.Expiration.Days) {
+                return true;
+            }
+
+            if (rule.Expiration.Date && rule.Expiration.Date < currentDate) {
+                return true;
+            }
+            // Expiration.ExpiredObjectDeleteMarker rule's action does not apply
+            // since object is not a delete marker.
+            // AbortIncompleteMultipartUpload.DaysAfterInitiation rule's action does not apply
+            // since in-progress MPUs are beeing handled separetly prior to this checks.
+        }
+
+        if (rule.Transitions && rule.Transitions.length > 0
+        && this._supportedRules.includes('transitions')) {
+            return rule.Transitions.some(t => {
+                if (t.Days !== undefined && daysSinceInitiated >= t.Days) {
+                    return true;
+                }
+                if (t.Date && t.Date < currentDate) {
+                    return true;
+                }
+                return false;
+            });
+        }
+
+        return false;
+    }
+
+    /**
+     * Check if entity (object or version) is eligible for expiration or transition based on its date.
+     * This function was introduced to avoid having to go further into processing
+     * (call getObjectTagging, headObject...) if the entity was not eligible based on its date.
+     * @param {array} rules - array of bucket lifecycle rules
+     * @param {object} entity - object or object version
+     * NOTE: entity is not an in-progress MPU or a delete marker.
+     * @param {string} versioningStatus - 'Enabled', 'Suspended', or 'Disabled'
+     * @return {boolean} true if eligible - false otherwise.
+     */
+    _isEntityEligible(rules, entity, versioningStatus) {
+        const currentDate = this._lifecycleDateTime.getCurrentDate();
+        const daysSinceInitiated = this._lifecycleDateTime.findDaysSince(
+            new Date(entity.LastModified)
+        );
+        const { staleDate } = entity;
+        const daysSinceStaled = staleDate ?
+            this._lifecycleDateTime.findDaysSince(new Date(staleDate)) : null;
+
+        // Always eligible if object is a current version delete marker because
+        // it requires extra s3 call (list versions).
+        if (entity.IsLatest && this._isDeleteMarker(entity)) {
```

Comment:
could move this to the top of the function as it doesn't use any of the constants defined.